### PR TITLE
Use a WC_Install-like approach to initialize WC Admin data

### LINKED ISSUE #532: Merging to WC Core: Use WC_Install approach for table updates, etc.
Core WooCommerce has a dedicated class for table updates (e.g. during activation/install). We should probably use it before/when wc-admin is merged to WooCommerce core.

See https://github.com/woocommerce/woocommerce/blob/14aea98d49b5afaa1b5f08c0cc1a1695392b089e/includes/class-wc-install.php

FWIW: Our existing approach (plugin activation hooks) is awesome for single sites, but not so great for network installs. See https://core.trac.wordpress.org/ticket/14170#comment:68 . Not sure if network installs are a requirement.

### PR DESCRIPTION
**Pull request by jeffstieler**
Fixes #532.

This PR introduces two new classes: `WC_Admin_Install` and `WC_Admin_Reports_Sync` that handle the initial table creation and data population as well as provide a home for database upgrade routine logic (that more closely matches WC core, making merge easier).

### Detailed test instructions:

- Use WCCOM dataset for best test results
- Disable wc-admin plugin
- Delete wc-admin lookup tables
- Activate plugin
- Verify that report table generation actions are queued and being processing
- Verify that "rebuild reports" button is under Status > Tools and does function properly

### CODE REVIEW

**Review by psealock — COMMENTED** [review]
With the plugin deactivated, I removed just a few tables:

```
MariaDB [tangaroa]> drop table wp_wc_admin_note_actions, wp_wc_customer_lookup, wp_wc_order_stats;
```
Activating or running the tool doesn't recreate them. So I drop the rest of the `wp_wc_*` tables and I get the following errors on activation and tool use.

screen shot 2019-02-11 at 2 35 58 pm

I'll leave my environment as it is for now if you need to know more.

EDIT: deactivating Woo, then wc-admin and reacitvating restored my site, but now I get strange data for Feb, which has no orders. I'm not sure if its related though.

screen shot 2019-02-11 at 2 21 53 pm

File: includes/class-wc-admin-install.php [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+	 * Return a list of tables. Used to make sure all WC Admin tables are dropped
+	 * when uninstalling the plugin in a single site or multi site environment.
+	 *
+	 * @return array WC tables.
+	 */
+	public static function get_tables() {
+		global $wpdb;
+
+		return array(
+			// @todo Will this work on multisite?
```
Comment (rrennick) [inline includes/class-wc-admin-install.php]: Yes, that works on multisite.

**Comment by rrennick** [comment]
@psealock try running the following SQL to trigger the schema update

`update wp_options set option_value = '0.5.0' where option_name = 'wc_admin_version';`

This PR worked for me & fixed #1509.

**Comment by jeffstieler** [comment]
> With the plugin deactivated, I removed just a few tables:
> ...
> Activating or running the tool doesn't recreate them.

@psealock - to @rrennick's point, the `create_tables()` method is only invoked if the WC admin version option value is less than the current WC admin version. If you don't delete/modify that value, the tables won't be recreated. Are you wanting to see logic that would rebuild the tables at any time (other than initial activation / upgrade)?

Thanks for the review @rrennick!

**Review by timmyc — APPROVED** [review]
Verified this is working against the wccom data set too 👍 Just a couple of minor comments but approving as-is.

File: includes/class-wc-admin-reports-sync.php [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+		add_action( self::ORDERS_LOOKUP_BATCH_INIT, array( __CLASS__, 'orders_lookup_batch_init' ) );
+		add_action( self::SINGLE_ORDER_ACTION, array( __CLASS__, 'orders_lookup_process_order' ) );
+	}
+
+	/**
+	 * Regenerate data for reports.
+	 */
+	public static function regenerate_report_data() {
+		// Add registered customers to the lookup table before updating order stats
+		// so that the orders can be associated with the `customer_id` column.
```
[the remark this answers is not in the document]
Comment (psealock) [reply includes/class-wc-admin-reports-sync.php]: The plumbing is already here and our Tracks events are getting rejected anyways, so we can add an event here.

**Comment by psealock** [comment]
>Are you wanting to see logic that would rebuild the tables at any time (other than initial activation / upgrade)?

Not particularly, I was trying to recreate a state of no relevant tables. Seems like its working as intended though.

File: includes/class-wc-admin-install.php [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+		set_transient( 'wc_admin_installing', 'yes', MINUTE_IN_SECONDS * 10 );
+		wc_maybe_define_constant( 'WC_ADMIN_INSTALLING', true );
+
+		self::create_tables();
+		WC_Admin_Reports_Sync::regenerate_report_data();
+		self::update_wc_admin_version();
+
+		delete_transient( 'wc_admin_installing' );
+
+		do_action( 'woocommerce_flush_rewrite_rules' ); // ?
```
[the remark this answers is not in the document]
Comment (jeffstieler) [reply includes/class-wc-admin-install.php]: Yep - I'll see if everything works without it and remove if so.

File: includes/class-wc-admin-install.php [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+			"{$wpdb->prefix}wc_admin_notes",
+			"{$wpdb->prefix}wc_admin_note_actions",
+			"{$wpdb->prefix}wc_customer_lookup",
+		);
+	}
+
+	/**
+	 * Adds new tables.
+	 *
+	 * @param array $wc_tables List of WooCommerce tables.
```
[the remark this answers is not in the document]
Comment (jeffstieler) [reply includes/class-wc-admin-install.php]: Actually they are WC tables - we're adding the WC Admin tables to them.

**Pull request merged**